code/modules/mob/living/carbon/human/species_types/psyphoza.dm

```diff
@@ -0,0 +1,276 @@
+		return
+	ping_turf(get_turf(owner))
+	has_cooldown_timer = TRUE
+	addtimer(CALLBACK(src, .proc/finish_cooldown), cooldown + (sense_time * min(1, overlays.len / PSYCHIC_OVERLAY_UPPER)))
+
+/datum/action/item_action/organ_action/psychic_highlight/proc/finish_cooldown()
+	has_cooldown_timer = FALSE
+
+//Allows user to see images through walls
+/datum/action/item_action/organ_action/psychic_highlight/proc/toggle_eyes_fowards()
+	//Grab organs - we do this here becuase of fuckery :tm:
+	if(!eyes && istype(owner, /mob/living/carbon/human))
+		var/mob/living/carbon/human/H = owner
+		//eyes
+		eyes = eyes || locate(/obj/item/organ/eyes) in H.internal_organs
+		sight_flags = eyes?.sight_flags
+		//Register signal for losing our eyes
+		RegisterSignal(eyes, COMSIG_PARENT_QDELETING, .proc/handle_eyes)
+
+	//handle eyes - make them xray so we can see all the things
+	eyes?.sight_flags = SEE_MOBS | SEE_OBJS | SEE_TURFS
+	owner.update_sight()
+
+//Dims blind overlay - Lightens highlight layer
+/datum/action/item_action/organ_action/psychic_highlight/proc/dim_overlay()
+	//Blind layer
+	var/atom/movable/screen/fullscreen/blind/psychic/P = locate (/atom/movable/screen/fullscreen/blind/psychic) in owner.client?.screen
+	if(P)
+		//We change the color instead of alpha, otherwise we'd reveal our actual surroundings!
+		P.color = "#000"
+		animate(P, color = P.origin_color, time = sense_time, easing = QUAD_EASING, flags = EASE_IN)
+	//Highlight layer
+	var/atom/movable/screen/plane_master/psychic/B = locate (/atom/movable/screen/plane_master/psychic) in owner.client?.screen
+	if(B)
+		B.alpha = 255
+		animate(B, alpha = 0, time = sense_time, easing = QUAD_EASING, flags = EASE_IN)
+	//Setup timer to delete image
+	if(overlay_timer)
+		deltimer(overlay_timer)
+	overlay_timer = addtimer(CALLBACK(src, .proc/toggle_eyes_backwards), sense_time, TIMER_STOPPABLE)
+
+//Get a list of nearby things & run 'em through a typecache
+/datum/action/item_action/organ_action/psychic_highlight/proc/ping_turf(turf/T, size = sense_range)
+	if(istype(owner?.get_item_by_slot(ITEM_SLOT_HEAD), /obj/item/clothing/head/helmet))
+		return
+	toggle_eyes_fowards()
+	dim_overlay()
+	//Get nearby 'things' to see with sense
+	var/list/nearby = range(size, T)
+	nearby -= owner
+	//Go through the list and render whitelisted types
+	for(var/atom/C as() in nearby)
+		if(is_type_in_typecache(C, sense_blacklist))
+			continue
+		highlight_object(C)
+
+//Add overlay for psychic plane
+/datum/action/item_action/organ_action/psychic_highlight/proc/highlight_object(atom/target)
+	//Build image
+	var/image/M = new()
+	M.appearance = target.appearance
+	M.transform = target.transform
+	M.pixel_x = 0 //Reset pixel adjustments to avoid bug where overlays tower
+	M.pixel_y = 0
+	M.pixel_z = 0
+	M.pixel_w = 0
+	M.plane = PSYCHIC_PLANE //Draw overlay on this plane so we can use it as a mask
+	M.dir = target.dir
+	//Add overlay for highlighting
+	target.add_overlay(M)
+	overlays += list(target, M)
```

Comment:
Rather than adding an overlay, set the images loc var and add it to the client images variable (Could use override too).

Reply:
This causes issues with in-hand icons not being visible. I originally used the method you're describing then swapped off to fix the bug.

Reply:
Didn't reply to this after a couple days, I'm just going to mark this as resolved.

Reply:
Unresolving since I believe add_overlay calls build_appearance which was having quite a lot of (add_overlay is usually called pretty infrequently, but this will be calling it in significant amounts).

Reply:
hm, thats strange it breaks with in-hand icons if its using the appearance of another object

Reply:
The latter, it just didn't show in the first place. However, I don't think the last few tests have proven the performance to be particularly terrible.
Perhaps some more testing and we can figure out if it's fine as-is.

Reply:
Alright, I'll have to log on during a testmerge but it might be a couple of weeks due to exams